preview/Win7Msix/Win7MSIXInstaller/Protocol.cpp

```diff
@@ -52,13 +105,91 @@ HRESULT Protocol::ExecuteForAddRequest()
     return S_OK;
 }
 
+HRESULT Protocol::ExecuteForAddRequest()
+{
+    for (std::vector<ProtocolData>::iterator protocol = m_protocols.begin(); protocol != m_protocols.end(); ++protocol)
+    {
+        RETURN_IF_FAILED(ProcessProtocolForAdd(*protocol));
+    }
+
+    return S_OK;
+}
+
+HRESULT Protocol::ProcessProtocolForAdd(ProtocolData& protocol)
+{
+    RegistryKey protocolKey;
+    RETURN_IF_FAILED(m_classesKey.CreateSubKey(protocol.name.c_str(), KEY_WRITE, &protocolKey));
+
+    RETURN_IF_FAILED(protocolKey.SetValue(protocolValueName.c_str(), nullptr, 0, REG_SZ));
+    RETURN_IF_FAILED(protocolKey.SetStringValue(L"", protocol.displayName));
+
+    if (protocol.logo.c_str() != nullptr)
+    {
+        RegistryKey defaultIconKey;
+        RETURN_IF_FAILED(protocolKey.CreateSubKey(defaultIconKeyName.c_str(), KEY_WRITE, &defaultIconKey));
+        RETURN_IF_FAILED(defaultIconKey.SetStringValue(L"", protocol.logo.c_str()));
+    }
+
+    RegistryKey shellKey;
+    RETURN_IF_FAILED(protocolKey.CreateSubKey(shellKeyName.c_str(), KEY_WRITE, &shellKey));
+    RETURN_IF_FAILED(shellKey.SetStringValue(L"", openKeyName));
+
+    RegistryKey openKey;
+    RETURN_IF_FAILED(shellKey.CreateSubKey(openKeyName.c_str(), KEY_WRITE, &openKey));
+
+    RegistryKey commandKey;
+    RETURN_IF_FAILED(openKey.CreateSubKey(commandKeyName.c_str(), KEY_WRITE, &commandKey));
+
+    std::wstring command = m_msixRequest->GetPackageInfo()->GetExecutableFilePath();
+    if (protocol.parameters.c_str() != nullptr)
+    {
+        command += std::wstring(L" ") + protocol.parameters;
+    }
+    else
+    {
+        command += commandArgument;
+    }
+    RETURN_IF_FAILED(commandKey.SetStringValue(L"", command));
+
+    return S_OK;
+}
+
+HRESULT Protocol::ExecuteForRemoveRequest()
+{
+    for (std::vector<ProtocolData>::iterator protocol = m_protocols.begin(); protocol != m_protocols.end(); ++protocol)
+    {
+        RETURN_IF_FAILED(ProcessProtocolForRemove(*protocol));
+    }
+
+    return S_OK;
+}
+
+HRESULT Protocol::ProcessProtocolForRemove(ProtocolData& protocol)
+{
+    HRESULT hrDeleteKey = m_classesKey.DeleteTree(protocol.name.c_str());
+    if (FAILED(hrDeleteKey))

```

Comment:
Probably check if the protocol is associated to the application we try to uninstall (verify the value of `open/command` first before removing the key)